File: bindings/ocaml/lib/operator.ml
(*
   Licensed to the Apache Software Foundation (ASF) under one
   or more contributor license agreements.  See the NOTICE file
   distributed with this work for additional information
   regarding copyright ownership.  The ASF licenses this file
   to you under the Apache License, Version 2.0 (the
   "License")you may not use this file except in compliance
   with the License.  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing,
   software distributed under the License is distributed on an
   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
   KIND, either express or implied.  See the License for the
   specific language governing permissions and limitations
   under the License.
*)

(** Core operator functions *)
let new_operator = Opendal_core.operator
let list = Opendal_core.blocking_list
let lister = Opendal_core.blocking_lister
let stat = Opendal_core.blocking_stat
let is_exist = Opendal_core.blocking_is_exist
let create_dir = Opendal_core.blocking_create_dir
let read = Opendal_core.blocking_read
let reader = Opendal_core.blocking_reader
let write = Opendal_core.blocking_write
let writer = Opendal_core.blocking_writer
let copy = Opendal_core.blocking_copy
let rename = Opendal_core.blocking_rename
let delete = Opendal_core.blocking_delete
let remove = Opendal_core.blocking_remove
let remove_all = Opendal_core.blocking_remove_all
let check = Opendal_core.blocking_check
let info = Opendal_core.operator_info

(** Reader module for reading data *)
module Reader = struct
  let pread = Opendal_core.reader_pread
end

(** Writer module for writing data *)
module Writer = struct
  let write = Opendal_core.writer_write
  let close = Opendal_core.writer_close
end

(** Lister module for streaming directory listings *)
module Lister = struct
  let next = Opendal_core.lister_next
end

(** Metadata module for file/directory information *)
module Metadata = struct
  let is_file = Opendal_core.metadata_is_file
  let is_dir = Opendal_core.metadata_is_dir
  let content_length = Opendal_core.metadata_content_length
  let content_md5 = Opendal_core.metadata_content_md5
  let content_type = Opendal_core.metadata_content_type
  let content_disposition = Opendal_core.metadata_content_disposition
  let etag = Opendal_core.metadata_etag
  let last_modified = Opendal_core.metadata_last_modified
end

(** Entry module for directory entry information *)
module Entry = struct
  let path = Opendal_core.entry_path
  let name = Opendal_core.entry_name
  let metadata = Opendal_core.entry_metadata
end

(** OperatorInfo module for operator information *)
module OperatorInfo = struct
  let name = Opendal_core.operator_info_name
  let scheme = Opendal_core.operator_info_scheme
  let root = Opendal_core.operator_info_root
  let capability = Opendal_core.operator_info_capability
end

(** Capability module for checking supported operations *)
module Capability = struct
  let stat = Opendal_core.capability_stat
  let read = Opendal_core.capability_read
  let write = Opendal_core.capability_write
  let create_dir = Opendal_core.capability_create_dir
  let delete = Opendal_core.capability_delete
  let copy = Opendal_core.capability_copy
  let rename = Opendal_core.capability_rename
  let list = Opendal_core.capability_list
  let list_with_limit = Opendal_core.capability_list_with_limit
  let list_with_start_after = Opendal_core.capability_list_with_start_after
  let list_with_recursive = Opendal_core.capability_list_with_recursive
  let presign = Opendal_core.capability_presign
  let presign_read = Opendal_core.capability_presign_read
  let presign_stat = Opendal_core.capability_presign_stat
  let presign_write = Opendal_core.capability_presign_write
  let shared = Opendal_core.capability_shared
end

File: bindings/cpp/src/details/operator.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "details/operator.hpp"

#include "details/common.hpp"
#include "lib.rs.h"

namespace opendal::details {

rust::Box<opendal::ffi::Operator> Operator::create_by_config(
    std::string_view scheme,
    const std::unordered_map<std::string, std::string> &config) {
  auto rust_map = rust::Vec<ffi::HashMapValue>();
  rust_map.reserve(config.size());
  for (auto &[k, v] : config) {
    rust_map.push_back({rust_string(k), rust_string(v)});
  }

  return opendal::ffi::new_operator(rust_str(scheme), rust_map);
}

std::string Operator::read(std::string_view path) {
  auto rust_vec = operator_->read(rust_str(path));
  return {rust_vec.data(), rust_vec.data() + rust_vec.size()};
}

void Operator::write(std::string_view path, std::string_view data) {
  operator_->write(rust_str(path), rust_slice<const uint8_t>(data));
}

rust::Box<ffi::Reader> Operator::reader(std::string_view path) {
  return operator_->reader(rust_str(path));
}

bool Operator::exists(std::string_view path) {
  return operator_->exists(rust_str(path));
}

void Operator::create_dir(std::string_view path) {
  operator_->create_dir(rust_str(path));
}

void Operator::copy(std::string_view src, std::string_view dst) {
  operator_->copy(rust_str(src), rust_str(dst));
}

void Operator::rename(std::string_view src, std::string_view dst) {
  operator_->rename(rust_str(src), rust_str(dst));
}

void Operator::remove(std::string_view path) {
  operator_->remove(rust_str(path));
}

ffi::Metadata Operator::stat(std::string_view path) {
  return operator_->stat(rust_str(path));
}

rust::Vec<ffi::Entry> Operator::list(std::string_view path) {
  return operator_->list(rust_str(path));
}

rust::Box<ffi::Lister> Operator::lister(std::string_view path) {
  return operator_->lister(rust_str(path));
}

}  // namespace opendal::details
